Adds explaination of `flatMap` semantics (#86)
WICG/observable: README.md

- `finally()`
  - Like `Promise.finally()`, it takes a callback which gets fired after the
    observable completes in any way (`complete()`/`error()`)

> - `flatMap()`

Reviewer: I don't think this should appear both in this "exceptional" list, _and_ in the list below about iterator helpers. Seems like it should just exist in the list of iterator helpers, with a note that the semantics are a little different (as you provided below).

means event handlers that call `e.preventDefault()` will run too late. See the
[Concerns](#concerns) section which goes into more detail.

### flatMap semantics

`flatMap` generally behaves like `Iterator.prototype.flatMap`, however, since it's push-based,
there can be a temporal element. Given that, it behaves much like RxJS's `concatMap`, which is
one of the most useful operators from the library.

At a high-level, it subscribes to the source, and then maps to, and emits values from "inner
observables", one at a time, ensuring they're subscribed to in sequence.

Given the following example:

```ts
const result = source.flatMap((value) => getNextInnerObservable(value));
```

- When you subscribe to `result`, it subscribes to `source` immediately.
- Let there be a `queue` of values that is empty.

> - Let there be an `innerSignal` that is either `undefined` or an `AbortSignal`

Author: That's correct. I'll add something.

> - Let there be an `innerSignal` that is either `undefined` or an `AbortSignal`

Author: Actually, I do explain that below: `Pass the `innerSignal` to the subscribe for the inner observable.`

- Let there be an `isSourceComplete` that is `false`.
- When the `source` emits a value:
  - If `innerSignal` is `undefined`

>     - Call the mapping function with the the value.

Author: Adding detail.

means event handlers that call `e.preventDefault()` will run too late. See the
[Concerns](#concerns) section which goes into more detail.

### flatMap semantics

`flatMap` generally behaves like `Iterator.prototype.flatMap`, however, since it's push-based,
there can be a temporal element. Given that, it behaves much like RxJS's `concatMap`, which is

> one of the most useful operators from the library.

Reviewer: > which is one of the most useful operators from the library.

I think we can remove this.

At a high-level, it subscribes to the source, and then maps to, and emits values from "inner
observables", one at a time, ensuring they're subscribed to in sequence.

Given the following example:

```ts
const result = source.flatMap((value, index) =>

> 	getNextInnerObservable(value, index),

Reviewer: What's the trailing comma here?

Author: Probably aggressive formatting rules. I'll have a look.

);
```

- `flatMap` is a method on `Observable` that is called with a `mapping function`, which takes a

>   value from the source observable, and a zero-based counter (or "index") of that value, and

Reviewer: ```suggestion
  value from the source observable, and a zero-based counter (or `index`) of that value, and
```

Just to make this consistent with `mapping function`

  returns a value that can be converted to an observable with `Observable.from`. `flatMap` returns
  an obeservable we'll call `result`.
- When you subscribe to `result`, it subscribes to `source` immediately.
- Let there be a `queue` of values that is empty.
- Let there be an integer `current index` that is `0`.

> - Let there be an `innerSignal` that is either `undefined` or an `AbortSignal`.

Reviewer: initially....? Where does this come from? It comes from the subscribe() method being called on the Observable returned by flatMap(), right? Maybe we should mention that. After reading the rest I'm actually thinking that `innerSignal` is not the one passed into subscribe. So... I'm not sure :)

Author: It's a bit of state for the internal algorithm. If `innerSignal` is `undefined`, there's no current inner subscription. If `innerSignal` is an `AbortSignal`, we have an inner subscription, and we need to wait for it to complete before we start another. So we need to buffer the value.

Basically:

```ts
let innerSignal: AbortSignal | undefined;
```

> - Let there be an `isSourceComplete` that is `false`.

Reviewer: ```suggestion
- Let there be a boolean `isSourceComplete` that is `false`.
```

- When the `source` emits a `value`:
  - If `innerSignal` is `undefined`
    - Begin **"mapping step"**:
      - Copy the `current index` into an `index` variable.
      - Increment the `current index`.

>       - Call the `mapping function` with the the `value` and the `index`.

Reviewer: ```suggestion
      - Call the `mapping function` with `value` and `index`.
```

>       - Then pass the return value of the mapping function to `Observable.from()` to convert it to

Reviewer: ```suggestion
      - Then pass the return value of the `mapping function` to `Observable.from()` to convert it to
```

        "inner observable" if it's not already.

>       - Then create an `AbortSignal` that is dependent on the subscriber's and set `innerSignal`.

Reviewer: ```suggestion
      - Then create an `AbortSignal` that is [dependent](https://dom.spec.whatwg.org/#create-a-dependent-abort-signal) on the subscriber's and set `innerSignal`.
```

>       - Pass the `innerSignal` to the subscribe for the inner observable.

Reviewer: ```suggestion
      - Subscribe to the inner observable, passing `innerSignal`
```

      - Forward all values emitted by the inner observable to the `result` observer.

>       - If the inner observable completes

Reviewer: ```suggestion
      - If the inner observable completes;
```

assuming you meant to coalesce this and the following condition.

      - If there are values in the `queue`
        - Take the first one from the `queue` and return to the **"mapping step"**.
      - If the `queue` is empty
        - If `isSourceComplete` is `true`

>         - Complete `result`.

Reviewer: Sometimes you have a period and sometimes you don't. Can you edit this whole algorithm to be consistent one way or another?

Author: Maybe I can.
Maybe I can't

        - If `isSourceComplete` is `false`

>         - Wait

Reviewer: Hmm, what is this supposed to do? Certainly we're not blocking the main thread....

Author: No... Just "wait for another event"... do nothing, I guess.

      - If the inner observable errors
        - Forward the error to `result`.

>   - Otherwise, if `innerSignal` is `AbortSignal`

Reviewer: ```suggestion
  - Otherwise, if `innerSignal` is an `AbortSignal`
```
